Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp

```diff
@@ -404,14 +404,41 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
 }
 
 #if HAVE(COOKIE_CHANGE_LISTENER_API)
-void WebCookieJar::addChangeListener(const String& host, const WebCore::CookieChangeListener& listener)
+void WebCookieJar::addChangeListenerWithAccess(const URL& url, const URL& firstParty, WebCore::FrameIdentifier frameID, WebCore::PageIdentifier pageID, WebCore::ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, const WebCore::CookieChangeListener& listener)
 {
-    auto& listenersForHost = m_changeListeners.add(host, WeakHashSet<CookieChangeListener> { }).iterator->value;
-    auto addResult = listenersForHost.add(listener);
-    if (listenersForHost.computeSize() > 1 || !addResult.isNewEntry)
+    auto host = url.host().toString();
+
+    if (auto iter = m_changeListeners.find(host); iter != m_changeListeners.end()) {
+        auto listenersForHost = iter->value;
+        if (listenersForHost.computeSize() > 1 || listenersForHost.contains(listener))
```

Comment:
We probably need to drop `listenersForHost.computeSize() > 1` and do the access check in network process for each listener. 
`listenersForHost.computeSize() > 1` only means the web process has a listener which successfully registers subscription for the host, but the new listener's access does not necessarily match the existing listener's access (unless the two listeners have the same url, firstPartyURL, frameID, pageID and shouldRelaxThirdPartyCookieBlocking); so we cannot assume the new listener will register successfully from `listenersForHost.computeSize() > 1`.